In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import pickle
from pandas_profiling import ProfileReport
import numpy as np

In [2]:
df= pd.read_csv('Advertising.csv')
df

Unnamed: 0     TV  radio  newspaper  sales
0             1  230.1   37.8       69.2   22.1
1             2   44.5   39.3       45.1   10.4
2             3   17.2   45.9       69.3    9.3
3             4  151.5   41.3       58.5   18.5
4             5  180.8   10.8       58.4   12.9
..          ...    ...    ...        ...    ...
195         196   38.2    3.7       13.8    7.6
196         197   94.2    4.9        8.1    9.7
197         198  177.0    9.3        6.4   12.8
198         199  283.6   42.0       66.2   25.5
199         200  232.1    8.6        8.7   13.4

[200 rows x 5 columns]

In [3]:
df.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [7]:
pf =ProfileReport(df)
pf.to_widgets()

In [8]:
pf.to_file('Advertising_Reprot.html')

In [13]:
x = df[['TV']]
x

TV
0    230.1
1     44.5
2     17.2
3    151.5
4    180.8
..     ...
195   38.2
196   94.2
197  177.0
198  283.6
199  232.1

[200 rows x 1 columns]

In [14]:
y=df[['sales']]
y

sales
0     22.1
1     10.4
2      9.3
3     18.5
4     12.9
..     ...
195    7.6
196    9.7
197   12.8
198   25.5
199   13.4

[200 rows x 1 columns]

In [11]:
from sklearn.linear_model import LinearRegression

In [12]:
linear = LinearRegression()

In [15]:
linear.fit(x,y)

LinearRegression()

In [16]:
linear.intercept_ # c of eq -  y=mx+c

array([7.03259355])

In [17]:
linear.coef_ # m of eq -  y=mx+c

array([[0.04753664]])

In [18]:
#So now we can say linear equation is 
#sales = (0.04753664) TV + 7.03259355
# now saving model in pickel format

file ='Advertise_linear_reg.sav'
pickle.dump(linear,open(file,'wb'))

In [20]:
linear.predict([[45]]) # Predicting sales value for TV =45

array([[9.17174237]])

In [21]:
linear.predict([[230.1]])

array([[17.97077451]])

In [24]:
lst = [16,256,1000,500,45]

for i in lst:
    print("for TV = {} we can expect Sales = {}".format(i ,linear.predict([[i]])))

for TV = 16 we can expect Sales = [[7.7931798]]
for TV = 256 we can expect Sales = [[19.2019735]]
for TV = 1000 we can expect Sales = [[54.56923398]]
for TV = 500 we can expect Sales = [[30.80091377]]
for TV = 45 we can expect Sales = [[9.17174237]]


In [25]:
# loading saved model

saved_model = pickle.load(open(file,'rb'))
saved_model.predict([[45]])

array([[9.17174237]])

In [26]:
linear.score(x,y) # r sqaure score to find model accuracy 

0.611875050850071

# Multivariable linear regression model 

In [4]:
df.columns

Index(['Unnamed: 0', 'TV', 'radio', 'newspaper', 'sales'], dtype='object')

In [5]:
x = df[['TV', 'radio', 'newspaper']]
y = df.sales

In [6]:
x

TV  radio  newspaper
0    230.1   37.8       69.2
1     44.5   39.3       45.1
2     17.2   45.9       69.3
3    151.5   41.3       58.5
4    180.8   10.8       58.4
..     ...    ...        ...
195   38.2    3.7       13.8
196   94.2    4.9        8.1
197  177.0    9.3        6.4
198  283.6   42.0       66.2
199  232.1    8.6        8.7

[200 rows x 3 columns]

In [7]:
y

0      22.1
1      10.4
2       9.3
3      18.5
4      12.9
       ... 
195     7.6
196     9.7
197    12.8
198    25.5
199    13.4
Name: sales, Length: 200, dtype: float64

In [10]:
from sklearn.linear_model import LinearRegression
lm  = LinearRegression()
lm.fit(x,y)

LinearRegression()

In [11]:
lm.intercept_

2.9388893694594085

In [12]:
lm.coef_

array([ 0.04576465,  0.18853002, -0.00103749])

In [13]:
lm.predict([[180.8,10.8,58.4]])

array([13.18867186])

In [14]:
lm.predict([[17.2,45.9,69.3]])

array([12.30767078])

In [ ]:
# Confidence menas from total 100 number of records for how may records molde giving accruate reason

In [15]:
lm.score(x,y) # r sqaure score to find model accuracy 

0.8972106381789522

# Using multiple variables in model-> giving higher accruacy than single variable model 
# also if notice coefficients then News paper coeff is -0.00103749 which is hardly impacting the result in comparosion with other variables 
# so we can neglect this variable 

In [ ]:
R sqaure is not reliable for each case so will consider Adjusted R square

In [17]:
import statsmodels.formula.api as smf
lm = smf.ols(formula='sales ~ TV',data=df).fit() # ols - >Ordinary list squad
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Sat, 28 Aug 2021   Prob (F-statistic):           1.47e-42
Time:                        15:44:06   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0326      0.458     15.360      0.000       6.130       7.935
TV             0.0475      0.003     17.668      0.000       0.042       0.053
==============================================================================
Omnibus:                        0.531   Durbin-Watson:                   1.935
Prob(Omnibus):                  0.767   Jarque-Bera (JB):                0.669
Skew:                          -0.089   Prob(JB):                        0.716
Kurtosis:                       2.779   Cond. No.                         338.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [18]:
lm = smf.ols(formula='sales ~ TV + radio',data=df).fit() # ols - >Ordinary list squad
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     859.6
Date:                Sat, 28 Aug 2021   Prob (F-statistic):           4.83e-98
Time:                        15:44:33   Log-Likelihood:                -386.20
No. Observations:                 200   AIC:                             778.4
Df Residuals:                     197   BIC:                             788.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9211      0.294      9.919      0.000       2.340       3.502
TV             0.0458      0.001     32.909      0.000       0.043       0.048
radio          0.1880      0.008     23.382      0.000       0.172       0.204
==============================================================================
Omnibus:                       60.022   Durbin-Watson:                   2.081
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              148.679
Skew:                          -1.323   Prob(JB):                     5.19e-33
Kurtosis:                       6.292   Cond. No.                         425.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [19]:
lm = smf.ols(formula='sales ~ TV + radio+newspaper',data=df).fit() # ols - >Ordinary list squad
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sat, 28 Aug 2021   Prob (F-statistic):           1.58e-96
Time:                        15:44:49   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""